src/lib/mf.rs

```diff
@@ -509,16 +476,398 @@ fn rank_cnds<TokId: Clone + Copy + Debug + TryFrom<usize> + TryInto<usize> + Par
+        let mut table = Vec::new();
+        table.resize(states_len * terms_len, u64::max_value());
+        loop {
+            let mut chgd = false; // Has anything changed?
+            for i in 0..states_len {
+                let edges = sgraph.edges(StIdx::from(i));
+                for (&sym, &sym_st_idx) in edges.iter() {
+                    let d = match sym {
+                        Symbol::Nonterm(nt_idx) => sengen.min_sentence_cost(nt_idx),
+                        Symbol::Term(t_idx) => {
+                            let off = usize::from(i) * terms_len + usize::from(t_idx);
+                            if table[off] != 0 {
+                                table[off] = 0;
+                                chgd = true;
+                            }
+                            term_cost(t_idx)
+                        }
+                    };
+
+                    for j in 0..terms_len {
+                        let this_off = usize::from(i) * terms_len + usize::from(j);
+                        let other_off = usize::from(sym_st_idx) * terms_len + usize::from(j);
+
+                        if table[other_off] != u64::max_value()
+                           && table[other_off] + d < table[this_off]
+                        {
+                            table[this_off] = table[other_off] + d;
+                            chgd = true;
+                        }
+                    }
+                }
+            }
+            if !chgd {
+                break;
+            }
+        }
+
+        // The following parts determine the distance in the presence of reductions and gotos.
+
+        // rev_edges allows us to walk backwards over the stategraph
+        let mut rev_edges = Vec::with_capacity(states_len);
+        rev_edges.resize(states_len, HashSet::new());
+        for i in 0..states_len {
+            for (_, &sym_st_idx) in sgraph.edges(StIdx::from(i)).iter() {
+                rev_edges[usize::from(sym_st_idx)].insert(StIdx::from(i));
+            }
+        }
+
+        // goto_edges is a map from a core state ready for reduction (i.e. where the dot is at the
+        // end of the production and thus sym_off == prod.len()) represented as a tuple
+        // (state_index, production_index) to a set of state indexes. The latter represents all the
+        // states which after (possibly recursive and intertwined) reductions and gotos the parser
+        // might end up in.
+        //
+        // We calculate this for a state i, production p_idx, by iterating backwards over the
+        // stategraph (using rev_edges) finding all routes of length grm.prod(p_idx).len() which
+        // could lead back to i, and then storing their goto state indexes. Since further
+        // reductions and gotos could then happen, we then have to recursively search those goto
+        // state indexes from scratch.
+        //
+        // This loop is currently comically inefficient, but it is relatively straightforward.
```

Comment:
Is it just on my machine or does git manage to cram three characters into one in the word inefficient? Even worse since it's a fixed-width font.
```
inefficient
aaaaaaaaaaa
```

Reply:
I don't let websites pick their fonts, so I don't know ;)